Pull Request for #8: Test status checks as a branch protection rule

veli-m-andirin opened the pull request:
@jmshapir, could you please enable "Require status checks to pass before merging" as a branch protection rule for master? We have a failing test in this branch which will trigger when I attempt to merge this pull. Hopefully the merge won't go through! (It would be sad if it does, but the code changes from this branch may be fine to have in master.)

jmshapir commented:
> @jmshapir, could you please enable "Require status checks to pass before merging" as a branch protection rule for master? We have a failing test in this branch which will trigger when I attempt to merge this pull. Hopefully the merge won't go through! (It would be sad if it does, but the code changes from this branch may be fine to have in master.)

@veli-m-andirin see PDF
[New branch protection rule.pdf](https://github.com/jmshapir/TestWorkflow/files/8526429/New.branch.protection.rule.pdf)
and screenshot:
![image](https://user-images.githubusercontent.com/7840353/164350605-85bfa89a-8fe4-4ae8-b78f-78325aa48ccb.png)

Let me know how it goes, thanks!

veli-m-andirin commented:
@jmshapir, the first two experiments here were unsuccessful:

I created a branch called `Veli` per https://github.com/jmshapir/TestWorkflow/pull/9#issuecomment-1104602726 and did the following:

- Modified the status check in `issue8_branch_protection` to trigger when pushing to `Veli`. I merged this PR without any issues, the status checks were queued but they never began, I cancelled those after 25 minutes.
- I thought that maybe GitHub only prevents merges when checks fail during a PR, so I created a new branch called `issue8_run_on_commits` and modified the workflow to trigger during the PR. Before merging, the status checks were failing but the merge happened again.. 

I will play with these a bit more today in a private repository, see if there's something else we need to do when setting branch protection rules and report back.

veli-m-andirin commented:
@jmshapir, I made some more experiments in the public repo `veli-m-andirin/Test`. One important thing I learnt is that, in order for GitHub to protect a branch from a failing status check, we should provide the `name` of the status check which can be assigned through the `yml` files (See [this commit](https://github.com/veli-m-andirin/Test/pull/2/commits/85ce35607e9e2a038b89a55b14ddead094af60c8)). Please find a branch protection selection here where I also chose the name of the status check: [Branch protection rule.pdf](https://github.com/jmshapir/TestWorkflow/files/8530626/Branch.protection.rule.pdf)

In terms of when and how this works, here are the observations:

- If we enable the status checks "before" merging the pull request (either throughout the pull or by removing commented-out parts just before merging), the branch protection rule works as intended, i.e. it doesn't allow merging a branch with failing tests:
![status_checks_must_pass](https://user-images.githubusercontent.com/62054739/164451563-aa6edc19-322e-4e37-90bd-f3d947176512.JPG)
- If we trigger the status checks just when we merge a branch (see [this commit](https://github.com/veli-m-andirin/Test/pull/2/commits/a3ad8f7cc415b52c23eb828bd21824b88955ca7b)), then it doesn't work as intended. GitHub wants to see the result of the test before merging, but the test triggers only when you click on "merge" and initiate the merge process, so there is a gridlock.
![status_checks_on_pull](https://user-images.githubusercontent.com/62054739/164452122-d33210e5-b3a3-439a-b71e-6faa7fab2262.JPG)

Given these, I propose to do the following:

- We run automated workflows when we are ready to merge a pull, and before merging. To do so, we comment and uncomment workflows manually.
- We name our workflows, and protect the master from merges if these named workflows are failing. 

However, I don't think the second step is terribly important. If we are conscious enough to comment and uncomment workflows to do the tests reasonably, then we'd probably be reasonable enough to avoid merging in the presence of failing tests, too. 

Although I am not extremely happy with the manual aspect of this procedure, I see one benefit: Sometimes there are long pull requests, and it might be valuable to run tests maybe at the beginning or in the middle, too, to get a better sense of which changes actually cause failures. (For PRs with a lot of commits, running the tests only at the end might make it harder to figure out where the problems were introduced.) I am not sure about how to calculate the memory usage of tests, but running a single unix test at the beginning and at the end of PRs would be light at least in terms of the minutes used.
